### ISSUE TITLE
[sig-storage] tests are massively failing in upgrade tests

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#103822]
Examples of failing jobs:
https://testgrid.k8s.io/google-gce-upgrade#gce-stable1-latest-upgrade-cluster
https://testgrid.k8s.io/google-gce-upgrade#gce-stable1-latest-upgrade-master

/kind failing-test
/priority important-soon
/sig storage

**Comment by dobsonj** [comment]
I took a quick look at the logs here:
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-master/1417728454077452288

And there are a lot of failures to list v1beta1 resources in the attacher and provisioner logs:
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-master/1417728454077452288/artifacts/_sig-storage_CSI_Volumes/_Driver_csi-hostpath_/_Testpattern_Dynamic_PV_default_fs_volume-stress/multiple_pods_should_access_different_volumes_repeatedly_Slow_Serial_/csi-hostpath-attacher-0-csi-attacher.log
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-master/1417728454077452288/artifacts/_sig-storage_CSI_Volumes/_Driver_csi-hostpath_/_Testpattern_Dynamic_PV_default_fs_volume-stress/multiple_pods_should_access_different_volumes_repeatedly_Slow_Serial_/csi-hostpath-provisioner-0-csi-provisioner.log
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-master/1417728454077452288/artifacts/_sig-storage_CSI_Volumes/_Driver_pd_csi_storage_gke_io_Serial_/_Testpattern_Dynamic_PV_default_fs_provisioning/should_provision_storage_with_mount_options/csi-gce-pd-controller-0-csi-attacher.log
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-master/1417728454077452288/artifacts/_sig-storage_CSI_Volumes/_Driver_pd_csi_storage_gke_io_Serial_/_Testpattern_Dynamic_PV_default_fs_provisioning/should_provision_storage_with_mount_options/csi-gce-pd-controller-0-csi-provisioner.log

Examples:
```
Failed to list *v1beta1.VolumeAttachment: the server could not find the requested resource
Failed to list *v1beta1.CSINode: the server could not find the requested resource
```

According to https://kubernetes.io/docs/reference/using-api/deprecation-guide/#storage-resources-v122 :
"The storage.k8s.io/v1beta1 API version of CSIDriver, CSINode, StorageClass, and VolumeAttachment will no longer be served in v1.22."

**Comment by wojtek-t** [comment]
Hmm - shouldn't the upgrade also upgrade the CSI drivers? @liggitt

**Comment by dims** [comment]
cc @xing-yang @msau42

**Comment by msau42** [comment]
We'll need to cherry pick parts of:
https://github.com/kubernetes/kubernetes/pull/100637
https://github.com/kubernetes/kubernetes/pull/102591

to the release-1.21 branch

**Comment by msau42** [comment]
Also https://github.com/kubernetes/kubernetes/pull/102212

**Comment by wojtek-t** [comment]
@msau42 - thanks!

It looks much better now, but still imperfect.
In particular the pv-metrics related tests are failing. It seems to be similar to https://github.com/kubernetes/kubernetes/issues/102865

@mauriciopoppe @pohly @aojea - can you please take a look as you were involved in the original bug?
[ I guess it's a matter of cherrypicking something]

**Comment by msau42** [comment]
Also looks like the hostpath reconstruction tests are failing. I'll work on backporting #103937

**Comment by mauriciopoppe** [comment]
@aojea do you know which PRs in addition to #103806 should be backported? I found a few that were related #102050, #101960, #101895

**Comment by aojea** [comment]
> @aojea do you know which PRs in addition to #103806 should be backported? I found a few that were related #102050, #101960, #101895


All those PRs were already backported, there should not be much difference right now between master and 1.22 ... let me check next week some of the failures to understand the cause

**Comment by wojtek-t** [comment]
@aojea - the upgrade tests are from stable1 to latest, which I think is 1.21 not 1.22 now.

**Comment by aojea** [comment]
I don't know how these jobs works for upgrades, but checking one of the ocurrences https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-latest-upgrade-cluster/1421760579051196416/build-log.txt
it seems that first runs the e2e test binary from master version against a cluster with the version 1.21.
```
I0801 09:19:34.328] Aug  1 09:19:34.327: INFO: e2e test version: v1.23.0-alpha.0.55+5be21c50c269fc
I0801 09:19:34.364] Aug  1 09:19:34.364: INFO: kube-apiserver version: v1.21.4-rc.0.32+1a8c1a4d628b9f
```

then it runs the e2e test binary from version 1.21 against the cluster using the new version 1.23
```
I0801 09:58:43.049] Aug  1 09:58:43.048: INFO: e2e test version: v1.21.4-rc.0.32+1a8c1a4d628b9f
I0801 09:58:43.089] Aug  1 09:58:43.088: INFO: kube-apiserver version: v1.23.0-alpha.0.55+5be21c50c269fc
I0801 09:58:43.089] Aug  1 09:58:43.088: INFO: >>> kubeConfig: /workspace/.kube/config
I0801 09:58:43.138] Aug  1 09:58:43.134: INFO: Cluster IP family: ipv4
```

One of the failures is that the test tries to get the metrics from port 10252, that was the controller manager insecure port and was deprecated in 1.22 https://github.com/kubernetes/kubernetes/pull/96216, hence the 1.21 test wants to se the 10252 port for metrics, but it is no longer available in 1.23

Should the upgrade job use only the e2e test binary from ci?
cc @BenTheElder

**Comment by BenTheElder** [comment]
I don't recall for certain anymore, but i thought it was one.

**Comment by aojea** [comment]
I can't find the logic that runs the e2e tests, and where it decides which e2e.test version to use

**Comment by msau42** [comment]
It is intended for the upgrade job to run e2e from a previous release against a newer release, to test that the newer release is backwards compatible. Although I believe we usually only test 1 version skew, not 2. However, I'm not sure that makes a difference here, as I understand, we had to make a number of changes in 1.22 to get the metrics tests to pass. Are those changes able to handle the metrics port differences between 1.21 and 1.22?

**Comment by pohly** [comment]
> However, I'm not sure that makes a difference here, as I understand, we had to make a number of changes in 1.22 to get the metrics tests to pass. Are those changes able to handle the metrics port differences between 1.21 and 1.22?

Those changes were all in the e2e.test binary, not core Kubernetes. So the root problem then is that e2e.test from Kubernetes 1.21 makes an assumption ("insecure port is enabled") that is no longer true for Kubernetes 1.22. The tests in e2e.test 1.21 which depend on metrics grabber simply won't work. Can we disable them?

To fix this, we would have to add code in 1.21 which detects the Kubernetes version and then selects different methods of grabbing metrics data. No such code exists at the moment, so this is more work than simply backporting some existing commits from master.

**Comment by wojtek-t** [comment]
+1 to what Michelle wrote above - we run tests from newer release to ensure backward compatibility.

> To fix this, we would have to add code in 1.21 which detects the Kubernetes version and then selects different methods of grabbing metrics data. No such code exists at the moment, so this is more work than simply backporting some existing commits from master.

Why do we need a code that is detecting versions?
Based on https://github.com/kubernetes/kubernetes/issues/103822#issuecomment-890564981 I see that the problem is connected to using insecure port. Can't we just backport the change that switched to secure port instead? [IIRC we had both secure and insecure back then, no?]

In other words, my question is - is backporting of some of the (test-only) changes from head not going to solve the problem?

**Comment by aojea** [comment]
> In other words, my question is - is backporting of some of the (test-only) changes from head not going to solve the problem?

Is not a clean backport

**Comment by msau42** [comment]
The 1.21 test runs both against release-1.21 and later versions. So my understanding is that the test needs to be able to handle both secure and insecure port configs.

**Comment by pohly** [comment]
> The 1.21 test runs both against release-1.21 and later versions. So my understanding is that the test needs to be able to handle both secure and insecure port configs.

That's exactly the reason.

**Comment by wojtek-t** [comment]
> The 1.21 test runs both against release-1.21 and later versions. So my understanding is that the test needs to be able to handle both secure and insecure port configs.

Doesn't 1.21 expose both ports? [And just starting 1.22 we deprecated only the insecure]

**Comment by pohly** [comment]
> Doesn't 1.21 expose both ports?

Hmm, good point. I don't know.

**Comment by pohly** [comment]
I tried backporting. It worked for a local cluster, let's see what the CI says: https://github.com/kubernetes/kubernetes/pull/104328

**Comment by msau42** [comment]
It's still failing. Although now I wonder if it's because of https://github.com/kubernetes/kubernetes/pull/102801. We'll need to backport the fix for https://github.com/kubernetes/kubernetes/issues/105965
